Add is_perfect_graph function and tests using Strong Perfect Graph Theorem

sourabh-sudesh-paradeshi opened the pull request:
This PR adds the is_perfect_graph function to check whether an undirected graph is perfect, based on the Strong Perfect Graph Theorem (SPGT).

Returns True if neither the graph nor its complement contains an induced odd cycle of length ≥ 5.

Uses cycle_basis to detect induced odd holes and antiholes.

Includes a fast check using chordality and an optional max_cycle_length parameter.

Adds tests for complete graphs, odd/even cycles, and complements.

This complements NetworkX’s structural graph tools by enabling efficient perfectness checks.

rossbar commented on networkx/algorithms/perfect_graph.py:
```diff
@@ -0,0 +1,80 @@
+__all__ = ["is_perfect_graph"]
+
+
+def is_perfect_graph(G, *, max_cycle_length=None):
```
I'm inclined to remove the `max_cycle_length`. I understand the desire to add it for performance reasons, but technically `is_perfect_graph` is a misnomer when `max_cycle_length` is anything other than `None`.

I also generally am -1 on the pattern of warning when checking for many cycles, which could naturally go away if we got rid of this kwarg.

rossbar reviewed (commented):
An algorithm for checking whether graphs are perfect graphs strikes me as a good fit for NetworkX. Thanks for the proposal!

There are some high-level considerations to hammer out (see inline comment) and then some tweaks to the code to improve things (e.g. get rid of the `is True` 's, etc.) but overall I'm in favor of this approach.

sourabh-sudesh-paradeshi commented:
> An algorithm for checking whether graphs are perfect graphs strikes me as a good fit for NetworkX. Thanks for the proposal!
> 
> There are some high-level considerations to hammer out (see inline comment) and then some tweaks to the code to improve things (e.g. get rid of the `is True` 's, etc.) but overall I'm in favor of this approach.

Thanks! I've removed max_cycle_length, cleaned up the logic, and squashed the commit. All related tests pass — the one failure seems unrelated. Let me know if anything else’s needed.

rossbar commented on networkx/algorithms/perfect_graph.py:
```diff
@@ -0,0 +1,37 @@
+def is_perfect_graph(G):
+    """Return True if G is a perfect graph, else False.
+
+    According to the Strong Perfect Graph Theorem, a graph is perfect
```
A reference should be included here, my vote would be to the wiki article.

rossbar commented on networkx/algorithms/perfect_graph.py:
```diff
@@ -0,0 +1,37 @@
+
+    According to the Strong Perfect Graph Theorem, a graph is perfect
+    if and only if neither it nor its complement contains an induced
+    odd cycle of length at least 5.
```
Similarly this docstring could include some information about the time complexity of the algorithm, since this is one of the big payoffs of the strong perfect graph theorem (i.e. the existence of a polynomial-time algorithm for checking whether a graph if perfect).

Finally, the docstring should be formatted according to the [numpydoc style guide](https://numpydoc.readthedocs.io/en/latest/format.html#sections) - i.e. there should be a `Parameters` section, etc. See other docstrings in the library for inspiration!

rossbar commented on networkx/algorithms/perfect_graph.py:
```diff
@@ -0,0 +1,37 @@
+    odd cycle of length at least 5.
+    """
+
+    def has_induced_odd_hole(graph):
+        nodes = list(graph.nodes)
+        n = len(nodes)
+        for i in range(n):
+            for j in range(i + 1, n):
+                try:
+                    paths = nx.all_simple_paths(graph, nodes[i], nodes[j], cutoff=n)
+                except nx.NetworkXNoPath:
+                    continue
+                for path in paths:
+                    if len(path) < 5 or len(path) % 2 == 0:
+                        continue
+                    if graph.has_edge(path[-1], path[0]):
+                        cycle_nodes = path
+                        subG = graph.subgraph(cycle_nodes)
+                        if subG.number_of_edges() == len(cycle_nodes):
+                            return True
+        return False
```
IMO this would be better as an external "private" function (i.e. `def _has_induced_odd_hole`) rather than a nested function. Doing so would have several advantages:
 1. Users don't have to worry about multiple scopes when reading the code (though arguably that's trivial in this case as there are no variables that cross scope boundaries)
 2. It allows the function to be independently tested
 3. It makes it easier to replace the function in the future if a better implementation is found.

rossbar commented on networkx/algorithms/perfect_graph.py:
```diff
@@ -0,0 +1,37 @@
+                            return True
+        return False
+
+    if has_induced_odd_hole(G):
+        return False
+    if has_induced_odd_hole(nx.complement(G)):
+        return False
```
Minor style nit, but might as well make this a single statement with an `or`

rossbar reviewed (commented):
It's not a blocker at all, but I'm curious if you've any insight on the path-based hole finding versus other "hole-finder" approaches. For example, what about using `simple_cycles` to enumerate all cycles, filtering on odd ones with len > 4, then checking the degree of all nodes in these filtered cycles is 2? Again not a blocker, but I'm curious about any additional research you may have done for the implementation!

Finally we'll have to make sure that this function is properly exported to the nx namespace (and think about the subpackage name) and included in the reference docs.

rossbar reviewed (commented):
This one's been stale for a while - are you still intending on working on this @sourabh-sudesh-paradeshi ? If so, I think an implementation based on `chordless_cycles` would generally be much more efficient than the proposed `has_induced_odd_hole`
